File: agent-authoring-python-kind/src/sretriage/agent.py
"""A Kubernetes SRE triage agent on Google ADK, hosted by the kagent Python runtime.

This file is the whole agent: a model, an instruction, one local tool and the MCP
tools kagent hands it. The A2A server, the streamed frames, the session and the task
the UI reads are all the runtime's work (kagent-adk), not this file's.
"""

import os

from google.adk import Agent
from google.adk.models.lite_llm import LiteLlm
from google.adk.telemetry.setup import maybe_set_otel_providers

from .gate import unhealthy
from .mcp import toolsets

# ADK creates spans for every model and tool call. They reach the collector kagent
# points at only if an OpenTelemetry provider is configured, which this does from the
# OTEL_* variables kagent injects.
os.environ.setdefault("OTEL_SERVICE_NAME", "sre-python")
maybe_set_otel_providers()

INSTRUCTION = """\
You are a Kubernetes SRE. When asked about a namespace, find the pods that are
unhealthy and explain why, from evidence.

Method: list the pods in the namespace with their phase and restart count. For each
pod call the unhealthy tool with its phase, restarts and, if it is Pending, how many
minutes it has been Pending; that tool is the rule, do not apply your own. For each
pod it marks unhealthy read the evidence that explains it: the pod description, the
last lines of its log, and the events for it. A container that never started has no
log, so use the events. Do not guess a cause you have not read.

Report format, plain text:
  <pod>  <state>  <one line cause>  <the one thing to check next>
one line per unhealthy pod, then one line naming the healthy pods. Keep it under
twenty lines. You have read-only tools and you do not change anything.
"""


def model() -> LiteLlm:
    """The model kagent asked for, read from MODEL_PROVIDER and MODEL_NAME.

    kagent injects both from the Agent record. LiteLLM takes "provider/model" and finds
    the provider's key in the environment (ANTHROPIC_API_KEY here).
    """
    provider = os.environ.get("MODEL_PROVIDER", "anthropic")
    name = os.environ.get("MODEL_NAME", "claude-haiku-4-5")
    return LiteLlm(model=f"{provider}/{name}")


root_agent = Agent(
    name="sre_triage",
    description="Kubernetes SRE triage. Reports which pods in a namespace are unhealthy, and why.",
    model=model(),
    instruction=INSTRUCTION,
    tools=[unhealthy, *toolsets()],
)

File: agent-authoring-go-kind/src/agent.go
package main

import (
	"context"
	"fmt"

	"github.com/go-logr/logr"
	"github.com/kagent-dev/kagent/go/adk/pkg/mcp"
	"github.com/kagent-dev/kagent/go/adk/pkg/models"
	"google.golang.org/adk/v2/agent"
	"google.golang.org/adk/v2/agent/llmagent"
	"google.golang.org/adk/v2/tool"
)

// instruction is the agent's system prompt. It is the same text every part of the
// series gives its agent, so the five agents can be compared on how they are built
// rather than on what they were told.
const instruction = `You are a Kubernetes SRE. When asked about a namespace, find the pods that are
unhealthy and explain why, from evidence.

A pod is unhealthy when it is not Running, or has restarted more than three
times, or has been Pending for more than five minutes. The tool "unhealthy"
applies that gate to one pod; use it rather than judging by eye.

Method: list the pods in the namespace, pick out the unhealthy ones, then for
each one read the evidence that explains it: the pod description, the last
lines of its log, and the events for it. A container that never started has
no log, so use the events. Do not guess a cause you have not read.

Report format, plain text:
  <pod>  <state>  <one line cause>  <the one thing to check next>
one line per unhealthy pod, then one line naming the healthy pods. Keep it under
twenty lines. You have read-only tools and you do not change anything.`

// newAgent builds the one LLM agent this program runs: the Anthropic model named by
// the environment, the MCP tools kagent approved, and the local gate tool.
//
// The MCP toolsets are created by the kagent library from the approved server list. The
// agent holds no credential for the tool server and does not know what is behind the
// URL; the gateway at that URL decides which tools this identity may see.
func newAgent(ctx context.Context, cfg Config, log logr.Logger) (agent.Agent, error) {
	model, err := models.NewAnthropicModelWithLogger(&models.AnthropicConfig{Model: cfg.Model}, log)
	if err != nil {
		return nil, fmt.Errorf("model: %w", err)
	}
	gate, err := newGateTool()
	if err != nil {
		return nil, fmt.Errorf("gate tool: %w", err)
	}
	toolsets := mcp.CreateToolsets(logr.NewContext(ctx, log), cfg.MCPServers, nil, false, nil)

	return llmagent.New(llmagent.Config{
		Name:        "sre_go",
		Description: "Kubernetes SRE triage. Reports which pods in a namespace are unhealthy, and why.",
		Instruction: instruction,
		Model:       model,
		Tools:       []tool.Tool{gate},
		Toolsets:    toolsets,
	})
}
